### PR TITLE
blas copy and axpy  to aten

### PR DESCRIPTION
**Pull request by aocsa**
Fixes #{issue number}

Follow-up PR: https://github.com/pytorch/pytorch/pull/50984

`copy` and `axpy` functions are ported to ATen. `THBlas_axpy` and `THBlas_copy` are removed.  

Looking forward your comments cc @ngimel, @mruberry

### CODE REVIEW

File: aten/src/ATen/native/CPUBlas.h [hunk 1]
```diff
@@ -94,4 +94,46 @@ void gemm(
...
+  }
+  axpy_stub(
+      kCPU, c10::CppTypeToScalarType<scalar_t>::value,
+      n, a, x, incx, y, incy);
+}
+
+void axpy(int64_t n, double a, double *x, int64_t incx, double *y, int64_t incy);
+void axpy(int64_t n, float a, float *x, int64_t incx, float *y, int64_t incy);
+void axpy(int64_t n, c10::complex<double> a, c10::complex<double> *x, int64_t incx, c10::complex<double> *y, int64_t incy);
+void axpy(int64_t n, c10::complex<float> a, c10::complex<float> *x, int64_t incx, c10::complex<float> *y, int64_t incy);
```
Comment (peterbell10) [inline aten/src/ATen/native/CPUBlas.h:116]: `x` should be `const *`.

File: aten/src/ATen/native/CPUBlas.h [hunk 2]
```diff
@@ -94,4 +94,46 @@ void gemm(
...
+  }
+  copy_stub(
+      kCPU, c10::CppTypeToScalarType<scalar_t>::value,
+      n, x, incx, y, incy);
+}
+
+void copy(int64_t n, double *x, int64_t incx, double *y, int64_t incy);
+void copy(int64_t n, float *x, int64_t incx, float *y, int64_t incy);
+void copy(int64_t n, c10::complex<double> *x, int64_t incx, c10::complex<double> *y, int64_t incy);
+void copy(int64_t n, c10::complex<float> *x, int64_t incx, c10::complex<float> *y, int64_t incy);
```
Comment (peterbell10) [inline aten/src/ATen/native/CPUBlas.h:137]: Same here: `const scalar_t *x`

**Review by peterbell10 — APPROVED** [review]

File: aten/src/ATen/native/cpu/BlasKernel.cpp [hunk 3]
```diff
@@ -189,9 +189,34 @@ void cpublas_gemm_impl(
       });
 }
 
+void cpublas_axpy_impl(at::ScalarType type, int64_t n, Scalar _a, const void *_x, int64_t incx, void *_y, int64_t incy){
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX(type, "cpublas_axpy_impl",
```
Comment (anjali411) [inline aten/src/ATen/native/cpu/BlasKernel.cpp:193]: `AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES`?

File: aten/src/ATen/native/cpu/BlasKernel.cpp [hunk 4]
```diff
@@ -189,9 +189,34 @@ void cpublas_gemm_impl(
...
+      auto x = static_cast<const scalar_t *>(_x);
+      auto y = static_cast<scalar_t *>(_y);
+      int64_t i;
+      for(i = 0; i < n; i++)
+        y[i*incy] += a*x[i*incx];
+    });
+}
+
+void cpublas_copy_impl(at::ScalarType type, int64_t n, const void *_x, int64_t incx, void *_y, int64_t incy){
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX(type, "cpublas_copy_impl",
```
Comment (anjali411) [inline aten/src/ATen/native/cpu/BlasKernel.cpp:205]: `AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES`

File: aten/src/ATen/native/CPUBlas.cpp [hunk 5]
```diff
@@ -250,4 +262,160 @@ void gemm(
...
+    incx = 1;
+    incy = 1;
+  }
+  #if AT_BUILD_WITH_BLAS()
+  if( (n <= INT_MAX) && (incx <= INT_MAX) && (incy <= INT_MAX) )
+  {
+    int i_n = (int)n;
+    int i_incx = (int)incx;
+    int i_incy = (int)incy;
+    daxpy_(&i_n, &a, x, &i_incx, y, &i_incy);
```
Comment (anjali411) [inline aten/src/ATen/native/CPUBlas.cpp:279]: this doesn't look correct. why does it always call into the function for `double`?

**Review by anjali411 — APPROVED** [review]
lgtm! just wondering if we should add the `axpy` and `copy` functions in `/home/chourdiaanjali/pytorch/aten/src/ATen/native/sparse/...` instead of `/home/chourdiaanjali/pytorch/aten/src/ATen/native/cpu/...`. @zou3519 what do you think?

**Comment by zou3519** [comment]
> lgtm! just wondering if we should add the `axpy` and `copy` functions in `/home/chourdiaanjali/pytorch/aten/src/ATen/native/sparse/...` instead of `/home/chourdiaanjali/pytorch/aten/src/ATen/native/cpu/...`. @zou3519 what do you think?

axpy and copy are general functions that operate on CPU that can be used for non-sparse applications. It just so happens in our case that we use them to accelerate our sparse tensor APIs. Because they are general purpose BLAS APIs I think we should keep them in native/cpu/BlasKernel.cpp

File: aten/src/ATen/native/CPUBlas.cpp [hunk 6]
```diff
@@ -250,4 +262,160 @@ void gemm(
...
+    return;
+  }
+  #endif 
+  axpy_stub(
```
Comment (ngimel) [inline aten/src/ATen/native/CPUBlas.cpp:283]: I have hard time believing that naive implementation in axpy_stub is better than y.add_(x, a) (or whatever the right syntax is for inplace add with alpha value set). Same for copy.

Comment (peterbell10) [reply aten/src/ATen/native/CPUBlas.cpp:283]: For small `n` it avoids overhead from tensor creation (for the slice), operator dispatch and `TensorIterator`.

Comment (ngimel) [reply aten/src/ATen/native/CPUBlas.cpp:283]: There's no tensor creation for inplace ops, and with blas-enabled builds it'll come here for large tensors, not small.

Comment (peterbell10) [reply aten/src/ATen/native/CPUBlas.cpp:283]: >There's no tensor creation for inplace ops

I'm talking about creating the view into the tensor in order to copy/axpy some subset of it. In all cases, that's how the `THBlas` `axpy`/`copy` operations were used.

> with blas-enabled builds it'll come here for large tensors, not small.

1. the naive code is primarily for non-blas types like integers.
2. you can also have small data with large strides and blas can't handle it

File: aten/src/ATen/native/CPUBlas.cpp [hunk 7]
```diff
@@ -250,4 +262,160 @@ void gemm(
...
+
+DEFINE_DISPATCH(copy_stub);
+
+void copy(int64_t n, const double *x, int64_t incx, double *y, int64_t incy) {
+  if(n == 1)
+  {
+    incx = 1;
+    incy = 1;
+  }
+  if( (n <= INT_MAX) && (incx <= INT_MAX) && (incy <= INT_MAX) ) {
```
[the remark this answers is not in the document]
Comment (aocsa) [reply aten/src/ATen/native/CPUBlas.cpp:359]: my bad, I fixed this

File: aten/src/ATen/native/CPUBlas.cpp [hunk 8]
```diff
@@ -250,4 +262,160 @@ void gemm(
...
+    int i_n = (int)n;
+    int i_incx = (int)incx;
+    int i_incy = (int)incy;
+    dcopy_(&i_n, x, &i_incx, y, &i_incy);
+  }
+  copy_stub(
```
[the remark this answers is not in the document]
Comment (aocsa) [reply aten/src/ATen/native/CPUBlas.cpp:365]: my bad, it shoudn't copy twice. I fixed in the next update.

File: aten/src/ATen/native/CPUBlas.cpp (hunk 6 above)
Comment (aocsa) [reply aten/src/ATen/native/CPUBlas.cpp:283]: > > There's no tensor creation for inplace ops
> 
> I'm talking about creating the view into the tensor in order to copy/axpy some subset of it. In all cases, that's how the `THBlas` `axpy`/`copy` operations were used.
> 
> > with blas-enabled builds it'll come here for large tensors, not small.
> 
> 1. the naive code is primarily for non-blas types like integers.
> 2. you can also have small data with large strides and blas can't handle it

I agree with Peter.

**Comment by aocsa** [comment]
Thanks @ngimel for the feedback.  I addressed the comments and fixed the  errors found in logic.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**